torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -431,84 +432,215 @@ c10::optional<at::Tensor> ComputeConstantFolding(Node* n, int opset_version) {
+    int allowzero = n->i(attr::allowzero);
+    if (allowzero == 1 && shape_has_zero) {
+      // Return here, because the denominator in shape_ratio calculation cannot
+      // be 0.
+      // TODO: Shall we handle the case when shape has -1 here?
+      return shape;
+    }
+  }
+
+  int minus_one_pos = -1;
+  for (int i = 0; i < shape_vector.size(); ++i) {
+    if (shape_vector[i].value() == -1) {
+      minus_one_pos = i;
+      break;
+    }
+  }
+
+  if (!shape_has_zero && minus_one_pos == -1) {
+    return shape;
+  }
+  std::vector<c10::ShapeSymbol> final_shape;
+  uint64_t shape_ratio = 1;
+  std::unordered_map<int64_t, int64_t> sym_map;
+  for (const c10::ShapeSymbol& input_shape : input_shape_vector) {
+    if (input_shape.is_static()) {
+      if (shape_ratio >=
+          std::numeric_limits<uint64_t>::max() / input_shape.static_size()) {
+        std::cerr
+            << "WARNING: ComputeShapeFromReshape(), shape_ratio overflows, skip shape inference."
+            << std::endl;
+        return c10::nullopt;
       } else {
-        shape_ratio /= static_cast<double>(input_shape[i]);
+        shape_ratio *= static_cast<uint64_t>(input_shape.static_size());
+      }
+    } else {
+      auto value = input_shape.value();
+      sym_map.emplace(value, 0).first->second += 1;
+    }
+  }
+  int shape_size = static_cast<int>(shape_vector.size());
+  for (const int i : c10::irange(shape_size)) {
+    if (i == minus_one_pos) {
+      continue;
+    }
+    c10::ShapeSymbol& target_shape = shape_vector[i];
+    if (target_shape.value() == 0) {
+      target_shape = input_shape_vector[i];
+    }
+    if (target_shape.is_static()) {
+      shape_ratio /= static_cast<uint64_t>(target_shape.static_size());
+    } else {
+      auto value = target_shape.value();
+      if (sym_map.find(value) == sym_map.end()) {
+        return c10::nullopt;
+      }
+      sym_map[value]--;
+      if (sym_map[value] == 0) {
+        sym_map.erase(value);
       }
     }
   }
 
+  // sym_map is used to match shape symbols between the input and shape.
+  // If there is a mismatch, the output shape cannot be estimated.
+  if (!sym_map.empty()) {
+    return c10::nullopt;
+  }
+
+  TORCH_INTERNAL_ASSERT(
+      minus_one_pos != -1,
+      "There are no examples for shape_has_zero = true && minus_one_pos == -1.");
+
   for (const auto i : c10::irange(minus_one_pos)) {
-    int64_t cur_shape = reshape[i] == 0 ? input_shape[i] : reshape[i];
+    c10::ShapeSymbol cur_shape(
+        shape_vector[i].value() == 0 ? input_shape_vector[i] : shape_vector[i]);
     final_shape.push_back(cur_shape);
   }
-  final_shape.push_back(static_cast<int64_t>(std::round(shape_ratio)));
-  for (auto i = minus_one_pos + 1; i < reshape_size; i++) {
-    int64_t cur_shape = reshape[i] == 0 ? input_shape[i] : reshape[i];
+  if (minus_one_pos != -1) {
+    final_shape.push_back(
+        ShapeSymbol::fromStaticSize(static_cast<int64_t>(shape_ratio)));
```

Comment:
Applying the following change as it broke internal CI:
```
stderr: xplat/caffe2/torch/csrc/jit/passes/onnx/shape_type_inference.cpp:635:9: error: use of undeclared identifier 'ShapeSymbol'; did you mean 'at::ShapeSymbol'?
        ShapeSymbol::fromStaticSize(static_cast<int64_t>(shape_ratio)));
        ^~~~~~~~~~~
        at::ShapeSymbol
xplat/caffe2/aten/src/ATen/core/jit_type.h:296:18: note: 'at::ShapeSymbol' declared here
struct TORCH_API ShapeSymbol {
                 ^
1 error generated.

```
```suggestion
        c10::ShapeSymbol::fromStaticSize(static_cast<int64_t>(shape_ratio)));
```